cmd/lotus-shed/hello.go

```diff
@@ -0,0 +1,63 @@
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"time"
+
+	cborutil "github.com/filecoin-project/go-cbor-util"
+	lcli "github.com/filecoin-project/lotus/cli"
+	"github.com/filecoin-project/lotus/node/hello"
+	"github.com/libp2p/go-libp2p"
+	inet "github.com/libp2p/go-libp2p-core/network"
+	"github.com/urfave/cli/v2"
+)
+
+var resultCh chan bool
+
+var helloCmd = &cli.Command{
+	Name:        "hello",
+	Description: "Get remote peer hello message by multiaddr",
+	ArgsUsage:   "[peerMultiaddr|minerActorAddress]",
+	Action: func(cctx *cli.Context) error {
+		ctx := lcli.ReqContext(cctx)
+		resultCh = make(chan bool, 1)
+		pis, err := lcli.AddrInfoFromArg(ctx, cctx)
+		if err != nil {
+			return err
+		}
+		h, err := libp2p.New()
+		if err != nil {
+			return err
+		}
+		err = h.Connect(ctx, pis[0])
+		if err != nil {
+			return err
+		}
+		h.SetStreamHandler(hello.ProtocolID, HandleStream)
```

Comment:
We should set the stream handler before connecting, if we don't do that, there's a small race that can occur:
* We connect to the peer
* They get list of our protocols, and don't see hello, so they don't send it
* We register the hello protocol, and wait for it
* It never comes because the peer thinks we don't support hello

Reply:
You are right , sorry for my carelessness